source/web/encryption.ts

```diff
@@ -0,0 +1,294 @@
+    // Output encrypted components
+    return {
+        method: EncryptionType.CBC,
+        auth: hmacHex,
+        iv: ivHex,
+        salt,
+        rounds,
+        content: encryptedContent
+    };
+}
+
+/**
+ * Encrypt text using AES-GCM
+ * @param text The text to encrypt
+ * @param keyDerivationInfo Key derivation information
+ * @param iv A buffer containing the IV
+ * @returns A promise that resolves with encrypted components
+ */
+export async function encryptGCM(
+    text: string,
+    keyDerivationInfo: DerivedKeyInfo,
+    iv: ArrayBuffer
+): Promise<EncryptedComponents> {
+    const crypto = getCrypto();
+    const { rounds } = keyDerivationInfo;
+    const ivArr = new Uint8Array(iv);
+    const ivHex = arrayBufferToHexString(iv);
+    const preparedContent = stringToArrayBuffer(text);
+    // Import keys (only primary)
+    // @ts-ignore
+    const importedKey = await crypto.subtle.importKey(
+        "raw",
+        keyDerivationInfo.key,
+        { name: ENC_ALGORITHM_GCM },
+        /* not extractable: */ false,
+        ["encrypt"]
+    );
+    // Encrypt content
+    const cipherBufferRaw = await crypto.subtle.encrypt(
+        {
+            name: ENC_ALGORITHM_GCM,
+            iv: ivArr,
+            additionalData: stringToArrayBuffer(`${ivHex}${keyDerivationInfo.salt}`)
+        },
+        importedKey,
+        preparedContent
+    );
+    // Extract auth tag
+    const tagLengthBytes = GCM_AUTH_TAG_LENGTH / 8;
+    const [cipherBuffer, tagBuffer] = extractAuthTag(cipherBufferRaw, tagLengthBytes);
+    const tag = arrayBufferToHexString(tagBuffer);
+    // Convert encrypted content to base64
+    const cipherArr = new Uint8Array(cipherBuffer);
+    const rawOutput = cipherArr.reduce((progress, byte) => {
+        return progress + String.fromCharCode(byte);
+    }, "");
+    const encryptedContent = btoa(rawOutput);
+    // Output encrypted components
+    return {
+        method: EncryptionType.GCM,
+        iv: ivHex,
+        salt: keyDerivationInfo.salt,
+        rounds,
+        content: encryptedContent,
+        auth: tag
+    };
+}
+
+function extractAuthTag(encryptedBuffer: ArrayBuffer, tagBytes: number): ArrayBuffer[] {
+    const shortedEnc = encryptedBuffer.slice(0, encryptedBuffer.byteLength - tagBytes);
+    const tagBuff = encryptedBuffer.slice(encryptedBuffer.byteLength - tagBytes);
+    return [shortedEnc, tagBuff];
+}
+
+/**
+ * IV generator
+ * @returns A newly generated IV
+ */
+export async function generateIV(): Promise<ArrayBuffer> {
+    const randArr = new Uint8Array(16);
+    getCrypto().getRandomValues(randArr);
+    return randArr.buffer;
+}
+
+/**
+ * Generate a random salt
+ * @param length The length of the string to generate
+ * @returns A promise that resolves with a salt (hex)
+ * @throws {Error} Rejects if length is invalid
+ */
+export async function generateSalt(length: number): Promise<string> {
+    if (length <= 0) {
```

Comment:
this function as well

Reply:
Same deal - node ones are sync, for now, but web ones are true async.